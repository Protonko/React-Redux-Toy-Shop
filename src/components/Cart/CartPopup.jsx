import React, {useRef} from 'react';
import PropTypes from 'prop-types';
import {Link} from 'react-router-dom';
import classNames from 'classnames';
import {useOutsideClick} from '../../hooks/useOutsideClick';
import {Scrollbars} from 'react-custom-scrollbars';
import {Product} from '../Product/Product';

export const CartPopup = ({setPopupVisible, removeFromCart, items}) => {
  const popup = useRef();
  const isItemsInCart = items.length;
  const classNamesButton = classNames('popup__button', 'button', {'button--disabled': !isItemsInCart});

  useOutsideClick(popup, () => setPopupVisible(false));

  return (
    <article className="popup drop-down" ref={popup}>
      <div className="popup__section popup__section--body">
        <Scrollbars>
          <div className="popup__scrollable-content">
            {isItemsInCart
              ? items.map((product, index) => (
                  <Product key={index} product={product} removeFromCart={removeFromCart} />
                )
              )
              : (
                <div className="popup__wrapper">
                  <p className="popup__text">
                    Your cart is empty.
                  </p>
                </div>
              )
            }
          </div>
        </Scrollbars>
      </div>

      <footer className="popup__section popup__section--footer">
        <Link to={'/order'} className={classNamesButton} onClick={() => setPopupVisible(false)}>
          Proceed to checkout
        </Link>
      </footer>
    </article>
  );
};

CartPopup.propTypes = {
  setPopupVisible: PropTypes.func,
  removeFromCart: PropTypes.func,
  items: PropTypes.array,
};
