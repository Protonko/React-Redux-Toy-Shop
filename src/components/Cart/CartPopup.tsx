import React, {FC, useRef} from 'react';
import {Link} from 'react-router-dom';
import classNames from 'classnames';
import {Scrollbars} from 'react-custom-scrollbars';
import {useOutsideClick} from 'hooks/useOutsideClick';
import {Product} from 'components/Product/Product';
import {IProduct} from 'models/interfaces';

interface ICartPopupProps {
  setPopupVisible: (popupVisible: boolean) => void,
  removeFromCart: (item: IProduct) => any, // TODO - убрать
  items: Array<IProduct>,
}

export const CartPopup: FC<ICartPopupProps> = ({
 setPopupVisible,
 removeFromCart,
 items,
}) => {
  const popup = useRef<HTMLElement>(null);
  const isItemsInCart: boolean = !!items.length;
  const classNamesButton: string =
    classNames('popup__button', 'button', {'button--disabled': !isItemsInCart});

  useOutsideClick<HTMLElement>(popup, () => setPopupVisible(false));

  const renderItem = (product: IProduct) => (
    <Product key={product.id} product={product} removeFromCart={removeFromCart} />
  );

  return (
    <article className="popup drop-down" ref={popup}>
      <div className="popup__section popup__section--body">
        <Scrollbars>
          <div className="popup__scrollable-content">
            {isItemsInCart
              ? items.map(renderItem)
              : (
                <div className="popup__wrapper">
                  <p className="popup__text">
                    Your cart is empty.
                  </p>
                </div>
              )
            }
          </div>
        </Scrollbars>
      </div>

      <footer className="popup__section popup__section--footer">
        <Link
          to={'/order'}
          className={classNamesButton}
          onClick={() => setPopupVisible(false)}
        >
          Proceed to checkout
        </Link>
      </footer>
    </article>
  );
};
