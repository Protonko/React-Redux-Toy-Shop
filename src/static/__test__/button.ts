export const ELEMENT_SELECTORS_BUTTON = {
  button: 'button',
  testButton: 'test-button',
  minusButton: 'button--count-minus',
}
