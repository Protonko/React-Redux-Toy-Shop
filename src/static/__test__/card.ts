// types
import {IProduct} from 'models/interfaces';

export const product: IProduct = {
  id: 0,
  category: 'category',
  title: 'title',
  price: 10,
  sale: 0.9,
  rating: 5,
  checked: true,
};

export const product1: IProduct = {
  id: 1,
  category: 'category 1',
  title: 'title 1',
  price: 1,
  sale: 0.5,
  rating: 3,
  checked: true,
}

export const products: Array<IProduct> = [product, product1];
export const selectedProducts: Array<IProduct> = [product];
