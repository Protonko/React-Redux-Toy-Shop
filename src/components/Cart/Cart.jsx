import React, {useState} from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import {ReactComponent as CartLogo} from '../../assets/icons/cart.svg';
import {CartPopup} from './CartPopup';

export const Cart = ({total, countProducts, items}) => {
  const [popupVisible, setPopupVisible] = useState(false);
  const classNamesButton = classNames(
    'cart__button',
    'button button--iconed',
    {'cart__button--counter': countProducts},
    {'cart__button--active': popupVisible}
  );

  return (
    <div className="cart">
      <ul className="cart__list list list--reset">
        <li className="cart__item">
          <div className="cart__price">
            <dl className="cart__price-list list list--reset">
              <dt className="cart__price-item cart__price-item--term">
                Total:
              </dt>
              <dd className="cart__price-item cart__price-item--description">
                ${total}
              </dd>
            </dl>
          </div>
        </li>
        <li className="cart__item">
          <button
            data-count={countProducts}
            className={classNamesButton}
            onClick={() => setPopupVisible(!popupVisible)}>
            <span className="cart__button-icon">
              <CartLogo/>
            </span>
          </button>

          {popupVisible && <CartPopup items={items} setPopupVisible={setPopupVisible}/>}
        </li>
      </ul>
    </div>
  );
};

Cart.propTypes = {
  total: PropTypes.number,
  countProducts: PropTypes.number,
  items: PropTypes.array,
};
