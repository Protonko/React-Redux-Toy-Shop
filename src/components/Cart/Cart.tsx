import React, {FC, useState} from 'react';
import {ReactComponent as CartLogo} from 'assets/icons/cart.svg';
import {TClassName} from 'models/types';
import {IProduct} from 'models/interfaces';
import {CartPopup} from './CartPopup';
import {ButtonIconed} from 'components/Common/Buttons/ButtonIconed';

interface IProps {
  total: number,
  countProducts: number,
  items: Array<IProduct>,
  removeFromCart: (item: IProduct) => void,
}

export const Cart: FC<IProps> = ({
  total,
  countProducts,
  items,
  removeFromCart,
}) => {
  const [popupVisible, setPopupVisible] = useState<boolean>(false);
  const classNamesButton: TClassName = [
    'cart__button',
    {'cart__button--counter': !!countProducts},
    {'cart__button--active': popupVisible}
  ];

  return (
    <div className="cart">
      <ul className="cart__list list list--reset">
        <li className="cart__item">
          <div className="cart__price">
            <dl className="cart__price-list list list--reset">
              <dt className="cart__price-item cart__price-item--term">
                Total:
              </dt>
              <dd className="cart__price-item cart__price-item--description">
                ${total}
              </dd>
            </dl>
          </div>
        </li>
        <li className="cart__item">
          <ButtonIconed
            data-count={countProducts}
            disabled={false}
            classNamesAdditional={classNamesButton}
            onClick={() => setPopupVisible(!popupVisible)}>
            <span className="cart__button-icon">
              <CartLogo />
            </span>
          </ButtonIconed>

          {popupVisible &&
            <CartPopup
              items={items}
              setPopupVisible={setPopupVisible}
              removeFromCart={removeFromCart}
            />
          }
        </li>
      </ul>
    </div>
  );
};
