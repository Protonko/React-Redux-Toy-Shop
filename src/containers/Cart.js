import {connect} from 'react-redux';
import {bindActionCreators} from 'redux';
import * as cartActions from 'store/actions/cart';
import {getTotalPrice} from 'store/selectors/cart';
import {Cart} from 'components/Cart/Cart';

const mapStateToProps = ({cart}) => ({
  items: cart.items,
  countProducts: cart.items.length,
  total: getTotalPrice(cart.items),
});

const mapDispatchToProps = dispatch => ({
  ...bindActionCreators(cartActions, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(Cart);
