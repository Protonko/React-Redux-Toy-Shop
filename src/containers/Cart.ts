// types
import {TInitialState} from 'store/reducers/cart';
import {IProduct} from 'models/interfaces';

import {connect} from 'react-redux';
import {bindActionCreators, Dispatch} from 'redux';
import * as cartActions from 'store/actions/cart';
import {getTotalPrice} from 'store/selectors/cart';
import {Cart} from 'components/Cart/Cart';

type TActions = {
  removeFromCart: (item: IProduct) => void,
}

const mapStateToProps = ({cart}: {cart: TInitialState}) => ({
  items: cart.items,
  countProducts: cart.items.length,
  total: getTotalPrice(cart.items),
});

const mapDispatchToProps = (dispatch: Dispatch) => ({
  ...bindActionCreators<TActions, TActions>(cartActions, dispatch),
});

export default connect(mapStateToProps, mapDispatchToProps)(Cart);
