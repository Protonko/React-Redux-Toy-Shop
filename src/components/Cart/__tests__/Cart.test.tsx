import React from 'react';
import {MemoryRouter} from 'react-router';
import {mount} from 'enzyme';
import {products} from 'static/__test__/card';
import {Cart, IProps} from 'components/Cart/Cart';

const mountCartComponent = (props: IProps, isRouter: boolean) => {
  if (isRouter) {
    return mount(<MemoryRouter><Cart {...props} /></MemoryRouter>);
  } else {
    return mount(<Cart {...props} />);
  }
}

describe('Cart', () => {
  let props: IProps;
  let popupVisible: boolean;

  beforeEach(() => {
    props = {
      total: 1,
      countProducts: 1,
      items: products,
      removeFromCart: () => {},
    };
    popupVisible = false;
  });

  it('Match snapshot', () => {
    const component = mountCartComponent(props, true);

    expect(component).toMatchSnapshot();
  });
});
