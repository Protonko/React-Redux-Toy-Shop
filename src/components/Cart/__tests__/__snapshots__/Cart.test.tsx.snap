// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Cart Match snapshot 1`] = `
<Router
  history={
    Object {
      "action": "POP",
      "block": [Function],
      "canGo": [Function],
      "createHref": [Function],
      "entries": Array [
        Object {
          "hash": "",
          "key": "m8op1i",
          "pathname": "/",
          "search": "",
          "state": undefined,
        },
      ],
      "go": [Function],
      "goBack": [Function],
      "goForward": [Function],
      "index": 0,
      "length": 1,
      "listen": [Function],
      "location": Object {
        "hash": "",
        "key": "m8op1i",
        "pathname": "/",
        "search": "",
        "state": undefined,
      },
      "push": [Function],
      "replace": [Function],
    }
  }
>
  <Component
    countProducts={1}
    getSelectedItems={[Function]}
    items={
      Array [
        Object {
          "category": "category",
          "checked": true,
          "id": 0,
          "price": 10,
          "rating": 5,
          "sale": 0.9,
          "title": "title",
        },
        Object {
          "category": "category 1",
          "checked": true,
          "id": 1,
          "price": 1,
          "rating": 3,
          "sale": 0.5,
          "title": "title 1",
        },
      ]
    }
    removeFromCart={[Function]}
    total={1}
  />
</Router>
`;
